### PR TITLE
Add -dependencyVersion CLI switch for install command

### PR DESCRIPTION
**Pull request by billboyletrintech**
Partial Fix for NuGet/Home#1806

### CODE REVIEW

**Comment by emgarten** [comment]
@billboyletrintech thanks for the PR!  Why is this a partial fix, what else is needed to fix the issue?

Also, this change would require tests. You can find the other nuget.exe tests under NuGet.Commandline.Tests

**Comment by billboyletrintech** [comment]
The scope of the issue this PR references is to add the commandline switch to install, update, and restore, as well as respecting the setting in nuget.config.  This PR only adds the switch to install as that's the use case I need in our continuous integration workflow.  I'll add the tests for install dependency resolution, didn't see that test project, sorry.

**Comment by emgarten** [comment]
Restore would be a much larger change to the resolver where this flag isn't supported today. Update should be very similar to install, does it look straight forward to add this there also?

**Comment by billboyletrintech** [comment]
It looked more complex because it appears that the current default behavior is to take the highest package with the same major/minor as the current downloaded package.  To me its a little unclear (without digging through the powershell implementation) what the correct behavior would be for all the different resolution options.  I'd rather someone more familiar with the code/logic/history does it.

**Comment by emgarten** [comment]
Update has ``-safe`` which applies to the package being updated. DependencyVersion may only be needed for dependencies. Ideally powershell and everything else would be consistent in how they treated this.

For install does the DependencyVersion actually change which top level package is being installed? Or is it always the latest?

**Comment by billboyletrintech** [comment]
No, my change only affects dependency resolution, top level is always latest if not specified.  It was actually a very simple fix, look at the diff.  The test was significantly more effort.

**Comment by serbrech** [comment]
Is there any plans around this?
I am hitting a similar use-case.
I want to update a package to the latest, but keep its dependencies to the lowest. that's not possible today it seems :(

File: src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs [hunk 1]
```diff
@@ -234,8 +237,23 @@ private CommandLineSourceRepositoryProvider GetSourceRepositoryProvider()
 
             var allowPrerelease = Prerelease || (version != null && version.IsPrerelease);
 
+            DependencyBehavior depBehavior;
+            if(String.IsNullOrWhiteSpace(DependencyVersion))
+            {
+                // default behavior
+                depBehavior = DependencyBehavior.Lowest;
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs]: @billboyletrintech I think this should read the value from the settings if it was not passed in.

**Review by emgarten — CHANGES_REQUESTED** [review]

**Comment by emgarten** [comment]
@serbrech would you add your scenario to https://github.com/NuGet/Home/issues/1806 to open a new issue if that one doesn't cover it?

**Comment by billboyletrintech** [comment]
You can pull and build my fork if you like, I don't expect them to merge this in anytime soon.

**Comment by emgarten** [comment]
:bell: @billboyletrintech are you able to make the changes needed to read this value from the config? 

I would like to make a decision on this PR soon, and to go into NuGet this needs to be a complete solution.

**Comment by serbrech** [comment]
@billboyletrintech I did not expand on my use case here, since it was described in the the orleans reference just below my message.

**Comment by billboyletrintech** [comment]
@emgarten ok I've added your requested changes

**Comment by mishra14** [comment]
@emgarten can you review/close this?

**Comment by emgarten** [comment]
@rrelyea

**Comment by erick-thompson** [comment]
Is this PR still under development? If not, I'd be happy to take it over and complete it. The issue that this PR addresses is a huge pain point.

**Comment by emgarten** [comment]
I think this PR is ready, my main concern here is that this is updating a legacy command that causes confusion for users. *Install* for example doesn't not handle packages with multiple target frameworks, it just picks one group at random and continues downloading packages. This is because no project framework is provided to make a decision against.

**Comment by erick-thompson** [comment]
I'm not what you mean about the confusion. The change is scoped to what
approach is taken when resolving dependencies. I wouldn't think anything
changes except the logic when a presented with a range. It doesn't change
anything about how project framework is used (or not).

**Review by emgarten — APPROVED** [review]

File: src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs [hunk 2]
```diff
@@ -209,6 +212,17 @@ private CommandLineSourceRepositoryProvider GetSourceRepositoryProvider()
             return new CommandLineSourceRepositoryProvider(SourceProvider);
         }
 
+        private DependencyBehavior GetSettingsDependencyBehavior(Configuration.ISettings settings)
+        {
+            DependencyBehavior ret;
+
+            string depSetting = SettingsUtility.GetConfigValue(settings, "dependencyVersion");
+            if (!Enum.TryParse<DependencyBehavior>(depSetting, out ret))
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs:220]: This should use ignore case when parsing the enum
https://msdn.microsoft.com/en-us/library/dd991317(v=vs.110).aspx

File: src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs [hunk 3]
```diff
@@ -233,9 +247,24 @@ private CommandLineSourceRepositoryProvider GetSourceRepositoryProvider()
             var primaryRepositories = packageSources.Select(sourceRepositoryProvider.CreateRepository);
 
             var allowPrerelease = Prerelease || (version != null && version.IsPrerelease);
+            
+            DependencyBehavior depBehavior = GetSettingsDependencyBehavior(Settings);
+
+            if (!String.IsNullOrWhiteSpace(DependencyVersion))
+            {
+                // if dependency version specified by command line, try to parse
+                if (!Enum.TryParse<DependencyBehavior>(DependencyVersion, out depBehavior))
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs:256]: Parse with ignore case set to true
https://msdn.microsoft.com/en-us/library/dd991317(v=vs.110).aspx

File: src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs [hunk 4]
```diff
@@ -233,9 +247,24 @@ private CommandLineSourceRepositoryProvider GetSourceRepositoryProvider()
             var primaryRepositories = packageSources.Select(sourceRepositoryProvider.CreateRepository);
 
             var allowPrerelease = Prerelease || (version != null && version.IsPrerelease);
+            
+            DependencyBehavior depBehavior = GetSettingsDependencyBehavior(Settings);
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs:251]: This does not need to be read if the command line flag is set, it should be moved to an else after that check below

File: src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs [hunk 5]
```diff
@@ -209,6 +212,17 @@ private CommandLineSourceRepositoryProvider GetSourceRepositoryProvider()
             return new CommandLineSourceRepositoryProvider(SourceProvider);
         }
 
+        private DependencyBehavior GetSettingsDependencyBehavior(Configuration.ISettings settings)
+        {
+            DependencyBehavior ret;
+
+            string depSetting = SettingsUtility.GetConfigValue(settings, "dependencyVersion");
+            if (!Enum.TryParse<DependencyBehavior>(depSetting, out ret))
+                ret = DependencyBehavior.Lowest; // if fails, set to default behavior
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs:221]: It seems like this one should fail also, you can create a common method that handles parsing the flag or failing with the new exception message.

File: src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx [hunk 6]
```diff
@@ -5336,4 +5336,10 @@ nuget locals global-packages -list</value>
   <data name="RestoreCommandRecursive" xml:space="preserve">
     <value>Restore all referenced projects for UWP and NETCore projects. This does not include packages.config projects.</value>
   </data>
+  <data name="InstallCommandDependencyVersion" xml:space="preserve">
+    <value>Overrides the default dependency resolution behavior</value>
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx:5340]: Add a `.` to the end, this is a complete sentence.

File: src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx [hunk 7]
```diff
@@ -5336,4 +5336,10 @@ nuget locals global-packages -list</value>
...
+  </data>
+  <data name="UnknownDependencyVersion" xml:space="preserve">
+    <value>Unknown Dependency Resolution Behavior for -dependencyVersion</value>
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx:5343]: Casing, and this could be simplified to
`Invalid value given for -DependencyVersion`

**Review by emgarten — CHANGES_REQUESTED** [review]
Some clean up is needed on this. Please also rebase it against the current dev branch and it will be considered for the next release

**Comment by WardenGnaw** [comment]
I'm interested in this PR going through and is blocking a couple work items for me. If @billboyletrintech does not make the changes in a couple days, can I take this over to get this going through?

**Comment by billboyletrintech** [comment]
I'm not going to have time to work on this I'm wrapping up a release at work.  Its been 8 months since my initial pull request and I've already made changes once.  Suddenly I need to make a bunch of changes within days?  Honestly I regret attempting to contribute.  Whoever wants to can feel free to take my work and 'finish' it.

**Comment by WardenGnaw** [comment]
@billboyletrintech. Sorry, I did not mean my comment to force a deadline on you. I thought that this PR may have been forgotten and was wondering if it was ok for someone else to resolve the last few comments if you were busy. 

Thank you for taking the initiative on fixing the issue.

**Comment by emgarten** [comment]
@billboyletrintech sorry you feel that way, I do appreciate the contribution.

@WardenGnaw feel free to fork and add to these changes, I can help you with with any additional feedback from there.

**Comment by emgarten** [comment]
Fixed with https://github.com/NuGet/NuGet.Client/commit/6f8d529312187fa1b6734da8d4ed9dc3442b916c

**Pull request closed without merging**